src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs

```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    public class AssemblyPart : ApplicationPart, IExportTypes
+    {
+        private readonly Assembly _assembly;
+
+        public AssemblyPart(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            _assembly = assembly;
+        }
+
+        public override string Name => _assembly.GetName().Name;
+
+        public IEnumerable<TypeInfo> Types => _assembly.DefinedTypes;
```

Comment:
`GetExportedTypes` would only be in our default implementation. I think that requiring a more advanced scenario that needs nested types or non-public types (which would be super weird) to do some extra work is reasonable, no?

Why are we interested in TypeInfo instead of Type? Does it offer any advantages? How do we reason about choosing one or the other?

Ultimately we can project Type to TypeInfo, so is that a problem here?